src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -400,19 +400,39 @@ public void OnCompleted(Func<object, Task> callback, object state)
 
         public void Flush()
         {
-            ProduceStartAndFireOnStarting().GetAwaiter().GetResult();
+            var produceStartTask = ProduceStartAndFireOnStarting();
+            if (produceStartTask.Status != TaskStatus.RanToCompletion)
+            {
+                produceStartTask.GetAwaiter().GetResult();
+            }
+
             SocketOutput.Write(_emptyData);
         }
 
-        public async Task FlushAsync(CancellationToken cancellationToken)
+        public Task FlushAsync(CancellationToken cancellationToken)
+        {
+            var produceStartTask = ProduceStartAndFireOnStarting();
+            if (produceStartTask.Status != TaskStatus.RanToCompletion)
+            {
+                return FlushAsyncAwaited(produceStartTask, cancellationToken);
+            }
+
+            return SocketOutput.WriteAsync(_emptyData, cancellationToken: cancellationToken);
+        }
+
+        public async Task FlushAsyncAwaited(Task produceStartTask, CancellationToken cancellationToken)
```

Comment:
So is this basically the same change you made to MessageBody where you used `ValueTask<int>`? I assume you're using normal tasks here since the Tasks here aren't generic.